src/firejail/preproc.c

```diff
@@ -21,12 +21,86 @@
 #include <sys/mount.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <sys/file.h>
 #include <dirent.h>
+#include <fcntl.h>
 
 static int tmpfs_mounted = 0;
 
+static void preproc_lock_file(const char *path, int *lockfd_ptr) {
+	assert(path != NULL);
+	assert(lockfd_ptr != NULL);
+
+	// Already locked
+	if (*lockfd_ptr != -1)
+		return;
+
+	long pid = (long)getpid();
+	if (arg_debug)
+		fprintf(stderr, "pid=%ld: locking %s ...\n", pid, path);
+
+	int lockfd = open(path, O_WRONLY | O_CREAT | O_CLOEXEC, S_IRUSR | S_IWUSR);
+	if (lockfd == -1) {
+		fprintf(stderr, "Error: cannot create a lockfile at %s\n", path);
+		errExit("open");
+	}
+
+	if (fchown(lockfd, 0, 0) == -1) {
+		fprintf(stderr, "Error: cannot chown root:root %s\n", path);
+		errExit("fchown");
+	}
+
+	if (flock(lockfd, LOCK_EX) == -1) {
+		fprintf(stderr, "Error: cannot lock %s\n", path);
+		errExit("flock");
+	}
+
+	*lockfd_ptr = lockfd;
+	if (arg_debug)
+		fprintf(stderr, "pid=%ld: locked %s\n", pid, path);
+}
+
+static void preproc_unlock_file(const char *path, int *lockfd_ptr) {
+	assert(path != NULL);
+	assert(lockfd_ptr != NULL);
+
+	int lockfd = *lockfd_ptr;
+	// Already unlocked
+	if (lockfd == -1)
+		return;
+
+	long pid = (long)getpid();
+	if (arg_debug)
+		fprintf(stderr, "pid=%ld: unlocking %s ...\n", pid, path);
+
+	if (flock(lockfd, LOCK_UN) == -1) {
+		fprintf(stderr, "Error: cannot unlock %s\n", path);
+		errExit("flock");
+	}
+
+	if (close(lockfd) == -1) {
+		fprintf(stderr, "Error: cannot close %s\n", path);
+		errExit("close");
+	}
+
+	*lockfd_ptr = -1;
+	if (arg_debug)
+		fprintf(stderr, "pid=%ld: unlocked %s\n", pid, path);
+}
+
+void preproc_lock_firejail_dir(int *lockfd_ptr) {
+	preproc_lock_file(RUN_DIRECTORY_LOCK_FILE, lockfd_ptr);
+}
+
+void preproc_unlock_firejail_dir(int *lockfd_ptr) {
+	preproc_unlock_file(RUN_DIRECTORY_LOCK_FILE, lockfd_ptr);
+}
+
 // build /run/firejail directory
-void preproc_build_firejail_dir(void) {
+void preproc_build_firejail_dir_unlocked(void) {
+	// Use atomic / otherwise timing hazard free operations
+	// to create enough of /run/firejail directory hierarchy
+	// to have a place for lock files
```

Comment:
Mmm, yeah, atomic is maybe a bit loaded / too strong word to use here. But effectively there is a sequence of mkdir() syscalls a firejail process executes / can execute. In case of multiple processes running in parallel, it does not matter which process / processes get to execute each one of these -> timing hazard free -> can be done without locking.

[One could argue that create_empty_dir_as_root() sort of does a bit too much and a bit too little: precheck with stat() to skip mkdir() is a "toctou" hazard and not checking type can lead to false positive, and mkdir() EEXISTS too could be due to something else than a directory. But in this context it does not really matter.] 

Whereas fs_remount_simple() & co perform sets of operations that have interdependence on each other and can easily be interfered by somebody else operating on the same paths / mount points -> locking is needed.

Reply:
> Mmm, yeah, atomic is maybe a bit loaded / too strong word to use here. But
> effectively there is a sequence of mkdir() syscalls a firejail process
> executes / can execute. In case of multiple processes running in parallel, it
> does not matter which process / processes get to execute each one of these ->
> timing hazard free -> can be done without locking.
> 
> [One could argue that create_empty_dir_as_root() sort of does a bit too much
> and a bit too little: precheck with stat() to skip mkdir() is a "toctou"
> hazard and not checking type can lead to false positive, and mkdir() EEXISTS
> too could be due to something else than a directory. But in this context it
> does not really matter.]

Agreed.

Misc: I started to refactor util.c some time ago to fix most toctou issues, but
it ended up depending on another WIP thing.

> Whereas fs_remount_simple() & co perform sets of operations that have
> interdependence on each other and can easily be interfered by somebody else
> operating on the same paths / mount points -> locking is needed.

Indeed; how about the following then?

```C
// build /run/firejail directory
//
// Note: This creates the base directory of the rundir lockfile;
// it should be called before preproc_lock_firejail_dir().
void preproc_build_firejail_dir_unlocked(void) {
```

Also, please squash/fixup the commits.

Reply:
Adjusted explanatory comments at preproc_build_firejail_dir_unlocked() as suggested.

Review commits squashed.